{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# COPILOT tutorial"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The data (sample name : col0) is available on GEO under record GSE152766. It was preprocessed to generate gene-by-cell matrices using scKB.\n",
    "\n",
    "#### Working directory must contain subdirectories,supp_data \n",
    "\n",
    "#### supp_data/ should contain the files, which are aviable on Github (github.com/Hsu-Che-Wei/COPILOT):\n",
    "\n",
    "    Protoplasting_DEgene_FC2_list.txt (from Denyer, et al. 2019)\n",
    "    color_scheme_at.RData (required if one choose to perform correlation-based annotation)\n",
    "    Root_bulk_arabidopsis_curated.RD (required if one choose to perform correlation-based annotation)\n",
    "\n",
    "#### The file 'Benfey_single_cell-Samples.csv' contains the meta data of samples, which is not provided. But one can find an example of such file in Supplementary Dataset 1 (Shahan & Hsu et al. 2020)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "rm(list=ls())\n",
    "# Set the working directory to where the folder named after the sample is located. \n",
    "# The folder contains spliced.mtx, unspliced.mtx, barcodes and gene id files, and json files produced by scKB that documents the sequencing stats. \n",
    "setwd(\"/scratch/AG_Ohler/CheWei/scKB\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load libraries\n",
    "suppressMessages(library(tidyverse))\n",
    "suppressMessages(library(COPILOT))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 3.6.1 (2019-07-05)\n",
       "Platform: x86_64-conda_cos6-linux-gnu (64-bit)\n",
       "Running under: CentOS Linux 7 (Core)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /fast/home/c/chsu/anaconda3/envs/seu314/lib/libopenblasp-r0.3.9.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.utf-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.utf-8        LC_COLLATE=en_US.utf-8    \n",
       " [5] LC_MONETARY=en_US.utf-8    LC_MESSAGES=en_US.utf-8   \n",
       " [7] LC_PAPER=en_US.utf-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.utf-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] COPILOT_0.1.0   forcats_0.5.0   stringr_1.4.0   dplyr_1.0.0    \n",
       " [5] purrr_0.3.4     readr_1.3.1     tidyr_1.1.0     tibble_3.0.1   \n",
       " [9] ggplot2_3.3.1   tidyverse_1.3.0\n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] readxl_1.3.1                uuid_0.1-4                 \n",
       "  [3] backports_1.1.7             DoubletFinder_2.0.3        \n",
       "  [5] plyr_1.8.6                  igraph_1.2.5               \n",
       "  [7] repr_1.1.0                  lazyeval_0.2.2             \n",
       "  [9] splines_3.6.1               BiocParallel_1.20.1        \n",
       " [11] listenv_0.8.0               GenomeInfoDb_1.22.1        \n",
       " [13] digest_0.6.25               htmltools_0.4.0            \n",
       " [15] fansi_0.4.1                 magrittr_1.5               \n",
       " [17] cluster_2.1.0               ROCR_1.0-11                \n",
       " [19] limma_3.42.2                globals_0.12.5             \n",
       " [21] modelr_0.1.8                matrixStats_0.56.0         \n",
       " [23] R.utils_2.9.2               colorspace_1.4-1           \n",
       " [25] blob_1.2.1                  rvest_0.3.5                \n",
       " [27] rappdirs_0.3.1              ggrepel_0.8.2              \n",
       " [29] haven_2.3.1                 crayon_1.3.4               \n",
       " [31] RCurl_1.98-1.2              jsonlite_1.6.1             \n",
       " [33] survival_3.1-12             zoo_1.8-8                  \n",
       " [35] ape_5.4                     glue_1.4.1                 \n",
       " [37] gtable_0.3.0                zlibbioc_1.32.0            \n",
       " [39] XVector_0.26.0              R2HTML_2.3.2               \n",
       " [41] leiden_0.3.3                DelayedArray_0.12.3        \n",
       " [43] DropletUtils_1.6.1          Rhdf5lib_1.8.0             \n",
       " [45] future.apply_1.5.0          SingleCellExperiment_1.8.0 \n",
       " [47] BiocGenerics_0.32.0         HDF5Array_1.14.4           \n",
       " [49] scales_1.1.1                DBI_1.1.0                  \n",
       " [51] edgeR_3.28.1                Rcpp_1.0.4.6               \n",
       " [53] viridisLite_0.3.0           reticulate_1.16            \n",
       " [55] dqrng_0.2.1                 rsvd_1.0.3                 \n",
       " [57] tsne_0.1-3                  stats4_3.6.1               \n",
       " [59] htmlwidgets_1.5.1           httr_1.4.1                 \n",
       " [61] RColorBrewer_1.1-2          ellipsis_0.3.1             \n",
       " [63] Seurat_3.1.5                ica_1.0-2                  \n",
       " [65] pkgconfig_2.0.3             R.methodsS3_1.8.0          \n",
       " [67] uwot_0.1.8                  dbplyr_1.4.4               \n",
       " [69] locfit_1.5-9.4              tidyselect_1.1.0           \n",
       " [71] rlang_0.4.6                 reshape2_1.4.4             \n",
       " [73] munsell_0.5.0               cellranger_1.1.0           \n",
       " [75] tools_3.6.1                 cli_2.0.2                  \n",
       " [77] generics_0.0.2              broom_0.5.6                \n",
       " [79] ggridges_0.5.2              evaluate_0.14              \n",
       " [81] fs_1.4.1                    fitdistrplus_1.1-1         \n",
       " [83] RANN_2.6.1                  pbapply_1.4-2              \n",
       " [85] future_1.17.0               nlme_3.1-148               \n",
       " [87] R.oo_1.23.0                 xml2_1.3.2                 \n",
       " [89] compiler_3.6.1              rstudioapi_0.11            \n",
       " [91] plotly_4.9.2.1              png_0.1-7                  \n",
       " [93] reprex_0.3.0                stringi_1.4.6              \n",
       " [95] lattice_0.20-41             IRdisplay_0.7.0            \n",
       " [97] Matrix_1.2-18               vctrs_0.3.1                \n",
       " [99] pillar_1.4.4                lifecycle_0.2.0            \n",
       "[101] lmtest_0.9-37               RcppAnnoy_0.0.16           \n",
       "[103] data.table_1.12.8           cowplot_1.0.0              \n",
       "[105] bitops_1.0-6                irlba_2.3.3                \n",
       "[107] patchwork_1.0.0             GenomicRanges_1.38.0       \n",
       "[109] R6_2.4.1                    gridExtra_2.3              \n",
       "[111] KernSmooth_2.23-17          IRanges_2.20.2             \n",
       "[113] codetools_0.2-16            MASS_7.3-51.6              \n",
       "[115] assertthat_0.2.1            rhdf5_2.30.1               \n",
       "[117] SummarizedExperiment_1.16.1 rjson_0.2.20               \n",
       "[119] withr_2.2.0                 sctransform_0.2.1          \n",
       "[121] S4Vectors_0.24.4            GenomeInfoDbData_1.2.2     \n",
       "[123] parallel_3.6.1              hms_0.5.3                  \n",
       "[125] grid_3.6.1                  IRkernel_1.1               \n",
       "[127] Rtsne_0.15                  pbdZMQ_0.3-3               \n",
       "[129] Biobase_2.46.0              lubridate_1.7.9            \n",
       "[131] base64enc_0.1-3            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load unwanted genes\n",
    "pp.genes <- as.character(read.table(\"./supp_data/Protoplasting_DEgene_FC2_list.txt\", header=F)$V1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare sample meta data for col0\n",
    "sample.name <- \"col0\"\n",
    "bscs <- read.csv(\"../../proj_sc/cbpsc/Benfey_single_cell-Samples.csv\", na.strings=c(\"\",\"NA\"))\n",
    "bscs <- bscs %>% select(c('sample','name','source','genotype','transgene','treatment','age','timepoint','rep','target_cells','date','seq_run')) %>% filter(sample==sample.name)\n",
    "bscs$date <- gsub('^([0-9]{4})([0-9]{2})([0-9]+)$', '\\\\1-\\\\2-\\\\3', bscs$date)\n",
    "bscs$target_cells <- prettyNum(bscs$target_cells, big.mark = ',')\n",
    "bscs <- t(bscs)\n",
    "sample.stats <- data.frame(stat = c('Sample','Name','Source','Genotype','Transgene','Treatment','Age','Timepoint','Rep','Target Cells','Date','Seq Run'), \n",
    "                    value = bscs[,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 12 Ã— 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>stat</th><th scope=col>value</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>sample</th><td>Sample      </td><td>col0      </td></tr>\n",
       "\t<tr><th scope=row>name</th><td>Name        </td><td>WT Col-0  </td></tr>\n",
       "\t<tr><th scope=row>source</th><td>Source      </td><td>Benfey lab</td></tr>\n",
       "\t<tr><th scope=row>genotype</th><td>Genotype    </td><td>WT Col-0  </td></tr>\n",
       "\t<tr><th scope=row>transgene</th><td>Transgene   </td><td>NA        </td></tr>\n",
       "\t<tr><th scope=row>treatment</th><td>Treatment   </td><td>untreated </td></tr>\n",
       "\t<tr><th scope=row>age</th><td>Age         </td><td>5_day     </td></tr>\n",
       "\t<tr><th scope=row>timepoint</th><td>Timepoint   </td><td>NA        </td></tr>\n",
       "\t<tr><th scope=row>rep</th><td>Rep         </td><td>NA        </td></tr>\n",
       "\t<tr><th scope=row>target_cells</th><td>Target Cells</td><td>5,000     </td></tr>\n",
       "\t<tr><th scope=row>date</th><td>Date        </td><td>NA        </td></tr>\n",
       "\t<tr><th scope=row>seq_run</th><td>Seq Run     </td><td>NA        </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 12 Ã— 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & stat & value\\\\\n",
       "  & <fct> & <fct>\\\\\n",
       "\\hline\n",
       "\tsample & Sample       & col0      \\\\\n",
       "\tname & Name         & WT Col-0  \\\\\n",
       "\tsource & Source       & Benfey lab\\\\\n",
       "\tgenotype & Genotype     & WT Col-0  \\\\\n",
       "\ttransgene & Transgene    & NA        \\\\\n",
       "\ttreatment & Treatment    & untreated \\\\\n",
       "\tage & Age          & 5\\_day     \\\\\n",
       "\ttimepoint & Timepoint    & NA        \\\\\n",
       "\trep & Rep          & NA        \\\\\n",
       "\ttarget\\_cells & Target Cells & 5,000     \\\\\n",
       "\tdate & Date         & NA        \\\\\n",
       "\tseq\\_run & Seq Run      & NA        \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 12 Ã— 2\n",
       "\n",
       "| <!--/--> | stat &lt;fct&gt; | value &lt;fct&gt; |\n",
       "|---|---|---|\n",
       "| sample | Sample       | col0       |\n",
       "| name | Name         | WT Col-0   |\n",
       "| source | Source       | Benfey lab |\n",
       "| genotype | Genotype     | WT Col-0   |\n",
       "| transgene | Transgene    | NA         |\n",
       "| treatment | Treatment    | untreated  |\n",
       "| age | Age          | 5_day      |\n",
       "| timepoint | Timepoint    | NA         |\n",
       "| rep | Rep          | NA         |\n",
       "| target_cells | Target Cells | 5,000      |\n",
       "| date | Date         | NA         |\n",
       "| seq_run | Seq Run      | NA         |\n",
       "\n"
      ],
      "text/plain": [
       "             stat         value     \n",
       "sample       Sample       col0      \n",
       "name         Name         WT Col-0  \n",
       "source       Source       Benfey lab\n",
       "genotype     Genotype     WT Col-0  \n",
       "transgene    Transgene    NA        \n",
       "treatment    Treatment    untreated \n",
       "age          Age          5_day     \n",
       "timepoint    Timepoint    NA        \n",
       "rep          Rep          NA        \n",
       "target_cells Target Cells 5,000     \n",
       "date         Date         NA        \n",
       "seq_run      Seq Run      NA        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sample.stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run copilot, please notice that do.annotation only supports root of Arabidopsis thaliana \n",
    "copilot(sample.name = \"col0\", species.name = \"Arabidopsis thaliana\", transcriptome.name = \"TAIR10\", sample.stats = sample.stats, mt.pattern = \"ATMG\", \n",
    "        mt.threshold = 5, cp.pattern = \"ATCG\", remove.doublet = TRUE, do.seurat = TRUE, do.annotation = TRUE, unwanted.genes = pp.genes, dir_to_color_scheme = \"./supp_data/color_scheme_at.RData\", \n",
    "        dir_to_bulk = \"./supp_data/Root_bulk_arabidopsis_curated.RD\", min.UMI.low.quality = 100, min.UMI.high.quality = 300)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# copilot\n",
    "\n",
    "# Description\n",
    "\n",
    "Single cell RNA-seq preprocessing tool for gene-by-cell matrices of UMI counts. It is recommended to use the raw spliced and unpliced counts matrices produced by scKB pipeline as the input of copilot.\n",
    "\n",
    "# Usage\n",
    "\n",
    "copilot(\n",
    "  sample.name,\n",
    "  spliced.mtx = NULL,\n",
    "  unspliced.mtx = NULL,\n",
    "  total.mtx = NULL,\n",
    "  filtered.mtx.output.dir = NULL,\n",
    "  species.name = \"Not Provided\",\n",
    "  transcriptome.name = \"Not Provided\",\n",
    "  sample.stats = NULL,\n",
    "  mt.pattern,\n",
    "  mt.threshold = 5,\n",
    "  cp.pattern = NULL,\n",
    "  top.percent = 1,\n",
    "  filtering.ratio = 1,\n",
    "  estimate.doublet.rate = TRUE,\n",
    "  doublet.rate = NULL,\n",
    "  remove.doublet = TRUE,\n",
    "  do.seurat = TRUE,\n",
    "  do.annotation = FALSE,\n",
    "  unwanted.genes = NULL,\n",
    "  HVG = FALSE,\n",
    "  HVGN = 200,\n",
    "  dir_to_bulk = NULL,\n",
    "  dir_to_color_scheme = NULL,\n",
    "  clustering_alg = 3,\n",
    "  res = 0.5,\n",
    "  min.UMI.low.quality = 100,\n",
    "  min.UMI.high.quality = 300,\n",
    "  legend.position = c(0.8, 0.8)\n",
    ")\n",
    "\n",
    "# Arguments\n",
    "\n",
    "sample.name:\n",
    "User defined sample name (character), which should be the same as the name of directory that contains spliced and unspliced matrices if you are following scKB pipeline to produce raw counts matrices.\n",
    "\n",
    "spliced.mtx:\n",
    "Gene by cell matrix of spliced counts, which should have column and row names, Default is NULL.\n",
    "\n",
    "unspliced.mtx:\n",
    "Gene by cell matrix of unspliced counts, which should have column and row names. Default is NULL.\n",
    "\n",
    "total.mtx:\n",
    "Gene by cell matrix of total counts, which should have column and row names. Default is NULL.\n",
    "\n",
    "filtered.mtx.output.dir:\n",
    "Output directory for quality filtered matrices. Default is NULL.\n",
    "\n",
    "species.name:\n",
    "Species name (character). Default is \"Not Provided\".\n",
    "\n",
    "transcriptome.name:\n",
    "Name of transcriptome annotation file. (e.g. TAIR10 for Arabidopsis). Default is \"Not Provided\".\n",
    "                                        \n",
    "sample.stats:\n",
    "Meta data of the sample in data.frame format. Default is NULL.\n",
    "                                        \n",
    "mt.pattern:\n",
    "Pattern of mitochondrial gene names/ids (character; e.g. \"ATMG\") or list of mitochondrial genes (character vector). This argument is required to run copilot.\n",
    "                                        \n",
    "mt.threshold:\n",
    "Threshold of mitochondrial expression percentage. Cell would be treated as dying cell if it has mitochodrial expression percentage higher than this threshold (numeric). Default is 5.\n",
    "                                        \n",
    "cp.pattern:\n",
    "Pattern of chloroplast gene names/ids (character; e.g. \"ATCG\") or list of chloroplast genes (character vector). Default is NULL.\n",
    "                                        \n",
    "top.percent:\n",
    "Percentage of cells that contain high numer of UMIs filtered (numeric). Default is 1.\n",
    "                                        \n",
    "filtering.ratio:\n",
    "Metric that controls the stringency of cell filtering (lenient: 1; strict:0; moderate: 0 < filtering.ratio < 1; numeric). Default is 1.\n",
    "                                        \n",
    "estimate.doublet.rate:\n",
    "Whether or not to estimate doublet rate according to 10X Genomics' estimation (boolean). Default is TRUE.\n",
    "\n",
    "doublet.rate:\n",
    "User specified doublet rate (numeric). Default is NULL.\n",
    "\n",
    "remove.doublet:\n",
    "Whether or not to remove doublets after quality filtering of gene and cell (boolean). Default is TRUE.\n",
    "\n",
    "do.seurat:\n",
    "Whether or not to perform normalization, PCA, UMAP and clustering using Seurat and output a Seurat object (boolean). Default is TRUE.\n",
    "                                        \n",
    "do.annotation:\n",
    "Whether or not to do annotation (boolean). COPILOT only supports annotation on root of Arabidopsis thaliana and Oryza sativa. Default is FALSE.\n",
    "                                        \n",
    "unwanted.genes:\n",
    "Gene IDs/names of unwanted genes (character vector, e.g. cell cycle related genes, organelle genes ... etc). Default is NULL.\n",
    "                                        \n",
    "HVG:\n",
    "Whether or not to select highly variable genes (boolean). Default is FALSE.\n",
    "                                        \n",
    "HVGN:\n",
    "Number of highly variable genes selected (numeric). Defalut is 200.\n",
    "                                        \n",
    "dir_to_bulk:\n",
    "Directory to reference expression profile for annotation. Default is NULL.\n",
    "\n",
    "dir_to_color_scheme:\n",
    "Directory to color scheme file for annotation. Default is NULL.\n",
    "                                        \n",
    "clustering_alg:\n",
    "Algorithm for clustering (1 = original Louvain algorithm; 2 = Louvain algorithm with multilevel refinement; 3 = SLM algorithm; 4 = Leiden algorithm, which requires the leidenalg python; numeric). Default is 3.\n",
    "\n",
    "res:\n",
    "Resolution used for clustering (numeric). Default is 0.5.\n",
    "\n",
    "min.UMI.low.quality:\n",
    "Minimum UMIs for a barcode to be considered as cell (numeric). Default is 100.\n",
    "\n",
    "min.UMI.high.quality:\n",
    "Minimum UMIs for a cell to be considered as high quality cell (numeric). Default is 300.\n",
    "\n",
    "legend.position:\n",
    "x y position of the legend on UMI histogram plot (numeric vector of length 2). Default is c(0.8,0.8)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
